### PR TITLE
feat(user): delete account data

### PR DESCRIPTION
**Pull request by naumovski-filip**
## Description

Notion card: https://www.notion.so/jwplayer/Viewer-delete-account-from-web-app-0cc99d39fe1d4a6e92d720afe9cee84b

### Steps completed:

According to our definition of done, I have completed the following steps:

- [X] Acceptance criteria met
- [ ] Unit tests added
- [ ] Docs updated (including config and env variables)
- [ ] Translations added
- [X] UX tested
- [ ] Browsers / platforms tested
- [ ] Rebased & ready to merge without conflicts
- [X] Reviewed own code

### CODE REVIEW

File: src/components/Account/Account.tsx
Comment (AntonLantukh) [inline src/components/Account/Account.tsx]:
```suggestion
      canDeleteAccount,
```

File: src/components/Account/Account.tsx
Comment (AntonLantukh) [inline src/components/Account/Account.tsx]: I would prefer doing it this way:  `canDeleteAccount && formSection({...`
And then we could filter the array using `filter(Boolean)` method.

File: src/components/DeleteAccountModal/DeleteAccountModal.tsx
Comment (AntonLantukh) [inline src/components/DeleteAccountModal/DeleteAccountModal.tsx]: Can we use `\n` in the translation?

File: src/i18n/locales/en_US/user.json
Comment (AntonLantukh) [inline src/i18n/locales/en_US/user.json]: I would suggest to group the keys this way:

```
 "confirmation_dialog": {
    "close": "Cancel",
    "confirm": "Yes"
  },
```

File: src/i18n/locales/en_US/user.json
Comment (AntonLantukh) [inline src/i18n/locales/en_US/user.json]: Can we replace numbers using subtitles?

File: src/services/inplayer.account.service.ts
Comment (AntonLantukh) [inline src/services/inplayer.account.service.ts]: Looks like part of the code we can reuse for several endpoints using a function. You can have a look on `getDataOrThrow` as an example.

File: src/containers/AccountModal/AccountModal.tsx
Comment (AntonLantukh) [inline src/containers/AccountModal/AccountModal.tsx]:
```suggestion
  const shouldHideBanner = view === 'delete-account'
```

File: src/containers/AccountModal/AccountModal.tsx
Comment (AntonLantukh) [inline src/containers/AccountModal/AccountModal.tsx]: Looks like we have already checked for a banner here

```suggestion
      {!shouldHideBanner && banner && <div className={styles.banner}><img src={banner} alt="" /></div>}
```

File: src/components/Dialog/Dialog.tsx
Comment (AntonLantukh) [inline src/components/Dialog/Dialog.tsx]:
```suggestion
      <div className={classNames(styles.dialog, size === 'large' && styles.largeDialog}>
```

File: src/components/Dialog/Dialog.module.scss
Comment (AntonLantukh) [inline src/components/Dialog/Dialog.module.scss:26]: Will it also work fine for mobile devices?

File: src/styles/_variables.scss
Comment (AntonLantukh) [inline src/styles/_variables.scss]: Maybe it is better to use a color names, what do you think (red-dark here)?
Btw, $red and $dark-danger-main look really close to each other, do we need both of them?

File: src/i18n/locales/en_US/user.json
Comment (naumovski-filip) [reply src/i18n/locales/en_US/user.json]: Not sure which subtitles to use here, open to any suggestions

File: src/components/DeleteAccountModal/DeleteAccountModal.tsx
Comment (naumovski-filip) [reply src/components/DeleteAccountModal/DeleteAccountModal.tsx]: I had issues with trying to use `\n` and break tags in the translation so decided to split them up in separate html elements

File: src/containers/AccountModal/AccountModal.tsx
Comment (naumovski-filip) [reply src/containers/AccountModal/AccountModal.tsx]: Implemented this as an array as there will be another case where hiding the banner is necessary (edit card details) which is also in review

File: src/styles/_variables.scss
Comment (naumovski-filip) [reply src/styles/_variables.scss]: I was going off the figma names when naming the colors and deciding on their hex values, I think it's best to keep them the same name as in figma so it's more obvious if they need to be reused in the future.

Comment (AntonLantukh) [reply src/styles/_variables.scss]: The problem is that it doesn't fit into the current paradigm where we have color names and not states colors correspond to. I would ask to change the figma names then to have the same approach.

File: src/i18n/locales/en_US/user.json
Comment (AntonLantukh) [reply src/i18n/locales/en_US/user.json]: @dbudzins should we translate all new keys to Spanish now?

File: src/components/Account/Account.tsx
Comment (AntonLantukh) [reply src/components/Account/Account.tsx]: I meant to add `filter(Boolean)` to the whole array of sections:

```
[
formSection({...,
canDeleteAccount && formSection({...
].filter(Boolean)
```

File: src/i18n/locales/en_US/user.json
Comment (AntonLantukh) [inline src/i18n/locales/en_US/user.json]: Do we need both `text_warning_1` and `warning_1` as well as `text_warning_2` and `warning_2`?

Comment (AntonLantukh) [inline src/i18n/locales/en_US/user.json]: Looks like we don't need this paragraph at all as the second one already includes this information

File: src/i18n/locales/en_US/user.json
Comment (AntonLantukh) [inline src/i18n/locales/en_US/user.json]: I would suggest to remove the first "However":

```
"warning_2": "All active subscriptions that were purchased through a web browser will be automatically canceled upon erasure of...."
```

File: public/locales/en/user.json
Comment (AntonLantukh) [inline public/locales/en/user.json]:
```suggestion
        "text_data_erasure": "Please note that if you choose to proceed with this process, all of your personal data will be erased from our system. This includes any personal information you provided during the registration process, as well as any location and device information that we gathered during the authentication process. Additionally, all other personal information stored in various parts of our database will be deleted as well. However, please be aware that due to some legislations we are required to keep certain personal data as part of transaction and subscription records.",
```

File: public/locales/en/user.json
Comment (AntonLantukh) [inline public/locales/en/user.json]:
```suggestion
        "text_revoked_access": "Once the erasure is complete, all of your active accesses will be revoked, and any active login sessions, including this one, will be terminated. Your account information will no longer be accessible to our staff in any capacity.",
```

File: public/locales/en/user.json
Comment (AntonLantukh) [inline public/locales/en/user.json]:
```suggestion
        "text_contacts": "If you have any further questions or concerns, please do not hesitate to contact our support team.",
```

File: public/locales/en/user.json
Comment (AntonLantukh) [inline public/locales/en/user.json]:
```suggestion
        "text_cant_be_undone": "Please be aware that this process cannot be undone, so proceed with caution.",
```

File: src/i18n/locales/en_US/user.json
Comment (budzworthy) [reply src/i18n/locales/en_US/user.json]: Yes, I'm asking around what the best way we can go about this is... We can start with a machine translation and then hopefully just fix up. I'm trying to get clarity on if we want to do this with each commit though...

File: src/components/Button/Button.tsx
Comment (ChristiaanScheermeijer) [inline src/components/Button/Button.tsx:11]: Shouldn't `danger` and `delete` be a color instead of a button variant?

File: src/components/Dialog/Dialog.tsx
Comment (ChristiaanScheermeijer) [inline src/components/Dialog/Dialog.tsx]: To make this expendable (and a bit cleaner), you could use the size as a key as well. 

```suggestion
      <div className={classNames(styles.dialog, styles[size])}>
```

In the CSS module you would have:

```scss
.small {
  max-width: 450px; // extracted from .dialog selector
}

.large {
  max-width: 768px;
}
```

File: src/containers/AccountModal/AccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/containers/AccountModal/AccountModal.tsx]: This function is now reactive and will be new for every render. This will cause problems in other modals because it is used in `useEffects` and `useCallback`, which are expected to be run only once. Therefor the function was wrapped with the `useEventCallback` hook.

Instead of checking the existence of the confirmation query param, I suggest always removing it without checking. Perhaps, you can even create a `removeQueryParams` utility that accepts multiple keys like so:

```ts
navigate(removeQueryParams(location, ['u', 'confirmation']));
```

Although, it might be weird that the AccountModal is aware of a query param that is set by a child component...

File: src/components/Button/Button.tsx
Comment (naumovski-filip) [reply src/components/Button/Button.tsx:11]: I put them as variants as they additionally have different styles for the outlines (both of them) and `outlined` is already a variant, so having them as colors seemed like it could potentially overlap styles with each other and wasn't that intuitive to me.

File: src/containers/AccountModal/AccountModal.tsx
Comment (naumovski-filip) [reply src/containers/AccountModal/AccountModal.tsx]: I'll go ahead with implementing a removeQueryParams with multiple parameters as arguments. Regarding AccountModal being aware of child query params, I do agree that the approach is not the best. I went with it because of the AccountModal/Dialog implementation and global state management solutions all seemed overkill. If you have a better idea I would be more than happy to reimplement it in this case

File: src/components/Account/Account.tsx
Comment (naumovski-filip) [reply src/components/Account/Account.tsx]: I can't seem to make this work, typescript is complaining about false still being a potential value. Could you try and suggest it as a commit exactly?

File: src/containers/AccountModal/AccountModal.tsx
Comment (ChristiaanScheermeijer) [reply src/containers/AccountModal/AccountModal.tsx]: It's better now since you won't need the `isConfirmation` variable. The `useEventCallback` is still missing, though.

I see two alternate solutions to refactor the `confirmation` query param.

1) By using local state in the `DeleteAccountModal` indicating it is the confirm step. We probably won't need to have a direct route to the confirmation state anyway
2) The second option is to create a separate `u` key for the confirmation step. You can still use the same component, but you must pass the password variable to the next modal. One solution for this is to use the same component instance and check the `u` parameter in the modal or pass the state as a prop like so;

```ts
switch (view) {
      //...
      case 'delete-account':
      case 'delete-account-confirmation':
        return <DeleteAccountModal confirmation={view === 'delete-account-confirmation'} />;
}
```

Comment (naumovski-filip) [reply src/containers/AccountModal/AccountModal.tsx]: Right, updated with `useEventCallback` in the latest commit. Regarding the alternate solutions, how would we make the `AccountModal`/`Dialog` components aware of the child's state then though? We need the info about which step the child component is at (in `AccountModal`) to handle the size of the `Dialog` component (at least with the current `Dialog` implementation).

In the second example, we'd still have to update the query params from `DeleteAccountModal`, and we'd still have to use the updated query parameter to be able to change the `Dialog` size.

File: src/containers/AccountModal/AccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/containers/AccountModal/AccountModal.tsx]: I think these deps are incorrect after fixing the merge conflicts. The `navigate` and `location` deps aren't needed anymore

File: src/containers/AccountModal/AccountModal.tsx
Comment (ChristiaanScheermeijer) [reply src/containers/AccountModal/AccountModal.tsx]: Yes, but the AccountModal is aware of the different views that it can render, so theoretically, you could use the same check as done for the `shouldHideBanner`;

```ts
  const shouldHideBanner = ['delete-account'].includes(view ?? '');
  const dialogSize = ['delete-account-confirmation'].includes(view ?? '') ? 'large' : 'small';  
```

The AccountModal only knows that there is a `delete-account-confirmation` view, but doesn't need to know more. 

From the `DeleteAccountModal` component (which should be a container because it uses a controller), you can navigate to the `?u=delete-account-confirmation` step after filling in the password.

File: public/locales/en/user.json
Comment (AntonLantukh) [inline public/locales/en/user.json]:
```suggestion
        "warning": "All active subscriptions that were purchased through a web browser will be automatically canceled upon erasure of personal data. However, if you purchased a subscription through iOS, Android, Roku, or Amazon apps, please be aware that it will not be canceled automatically. In this case, you will need to follow the usual subscription cancelation steps for each respective platform to manually cancel the subscription."
```

File: src/containers/AccountModal/AccountModal.tsx
Comment (AntonLantukh) [inline src/containers/AccountModal/AccountModal.tsx]: We could probably change it to `shouldHShowBanner` not to negate the expression always.

**Review by AntonLantukh — COMMENTED** [review]
Looks good to me! Thanks for addressing all the comments! Just waiting for the final decision on translations.

File: src/components/DeleteAccountModal/DeleteAccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/components/DeleteAccountModal/DeleteAccountModal.tsx:29]: When entering a wrong password this modal seems to be closed and reset. Is this clear enough for the user that the entered password was wrong?

File: src/components/DeleteAccountModal/DeleteAccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/components/DeleteAccountModal/DeleteAccountModal.tsx]: This is not needed anymore, right?

File: src/containers/AccountModal/AccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/containers/AccountModal/AccountModal.tsx]: This is also not needed anymore?

**Comment by budzworthy** [comment]
> Looks good to me! Thanks for addressing all the comments! Just waiting for the final decision on translations.

Since it's someone outside of engineering who is essentially volunteering, I'd suggest we don't put too much pressure onto them by making them a critical path for web app commits. Let's do the thing of opening 2 pr's (1 for logic with English and 1 for (spanish) translations) for now.

**Review by AntonLantukh — APPROVED** [review]

File: src/components/DeleteAccountModal/DeleteAccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/components/DeleteAccountModal/DeleteAccountModal.tsx]: Minor and opinionated, but perhaps sending the user back to the first step is more user-friendly instead of closing the modal? 

```suggestion
  const handleCancel = () => {
    navigate(removeQueryParam(location, 'u', { replace: true }));
  };
  
  const handleAlertClose = () => {
    // reset password (perhaps a state reset of the form is also needed)
    setEnteredPassword('');
    // go back one step
    navigate(addQueryParam(location, 'u', 'delete-account'), { replace: true });
  };

  if (deleteAccount.isError) {
    return <Alert open isSuccess={false} onClose={handleCancel} message={t('account.delete_account.error')} />;
  }
```

File: src/components/DeleteAccountModal/DeleteAccountModal.tsx
Comment (ChristiaanScheermeijer) [inline src/components/DeleteAccountModal/DeleteAccountModal.tsx]: Not sure if you want to be able to use the back button. I think it ends up in a mismatch between the `u` parameter and the presented modal because the confirmation modal is shown based on the `enteredPassword` state.

```suggestion
      navigate(addQueryParam(location, 'u', 'delete-account-confirmation', { replace: true }));
```

**Comment by ChristiaanScheermeijer** [comment]
Thanks for the changes and fixes @naumovski-filip! Nice work 😄 

I have one concern left regarding the history state and the modal. I think the URL needs to be replaced instead of pushed to prevent the user from going back to the state by using the browser's back/forward buttons. 

@dbudzins what do you think?

**Comment by naumovski-filip** [comment]
> Thanks for the changes and fixes @naumovski-filip! Nice work 😄
> 
> I have one concern left regarding the history state and the modal. I think the URL needs to be replaced instead of pushed to prevent the user from going back to the state by using the browser's back/forward buttons.
> 
> @dbudzins what do you think?

@ChristiaanScheermeijer Take a look at how I approached it in https://github.com/jwplayer/ott-web-app/pull/279/commits/eacd3b7575cc0d6cbf891c7ed7943c9b4b66ab76 and tell me what you think. I think I tested out every case, back button, redirect after error, writing the url without entering the password etc.

**Comment by budzworthy** [comment]
> > Thanks for the changes and fixes @naumovski-filip! Nice work smile
> > I have one concern left regarding the history state and the modal. I think the URL needs to be replaced instead of pushed to prevent the user from going back to the state by using the browser's back/forward buttons.
> > @dbudzins what do you think?
> 
> @ChristiaanScheermeijer Take a look at how I approached it in [eacd3b7](https://github.com/jwplayer/ott-web-app/commit/eacd3b7575cc0d6cbf891c7ed7943c9b4b66ab76) and tell me what you think. I think I tested out every case, back button, redirect after error, writing the url without entering the password etc.

I agree with Christiaan on this one. It's something that's bugged me with other modals in the app that need to be fixed and somewhere there's a backlog ticket for it. When a modal is shown and then closed, I shouldn't see it pop up again when I use the back button (it's not enough to just clear the modal state.) I think using `replace` achieves this.

**Review by darkoatanasovski — APPROVED** [review]

**Comment by AntonLantukh** [comment]
@naumovski-filip could you please create a PR with Spanish translations?

**Pull request merged**
